conduit/pipeline/pipeline.go

```diff
@@ -431,8 +467,250 @@ func addMetrics(block data.BlockData, importTime time.Duration) {
+				select {
+				case <-p.ctx.Done():
+					return
+				case blkOutChan <- pluginBlk:
+					waitTime := time.Since(selectStart)
+					totalFeedWait += waitTime
+					p.logger.Tracef("processor[%d] %s for round %d into blkOutChan after waiting %dms", idx, proc.Metadata().Name, lastRnd, totalFeedWait.Milliseconds())
+				}
+			}
+		}
+	}()
+}
+
+// exporterHandler handles the exporter's Receive method, updating the metadata's NextRound,
+// saving the metadata, and invoking the activated plugin callbacks. If the context's cancellation is received
+// before the exporter's Receive method has succeeded, cancel immediately.
+// However, after the exporter's Receive() method has succeeded we try every component at least once
+// but will abort if a failure occurs after a cancellation is received.
+func (p *pipelineImpl) exporterHandler(exporter exporters.Exporter, blkChan pluginChannel) {
+	p.wg.Add(1)
+	go func() {
+		defer p.wg.Done()
+		var lastError error
+		lastRound := uint64(0)
+		totalSelectWait := time.Duration(0)
+		totalExportWork := time.Duration(0)
+
+		defer func() {
+			p.logger.Debugf("exporter %s handler exiting. lastRnd=%d totalSelectWait=%dms, totalExportWork=%dms", exporter.Metadata().Name, lastRound, totalSelectWait.Milliseconds(), totalExportWork.Milliseconds())
+			if lastError != nil {
+				err := fmt.Errorf("exporter %s handler (%w) after round %d: %w", exporter.Metadata().Name, errExporterCause, lastRound, lastError)
+				p.cancelWithProblem(err)
+				p.logger.Error(err)
+			}
+		}()
+
+		for {
+			p.logger.Tracef("exporter handler waiting for block data. round=%d", lastRound)
+			selectStart := time.Now()
+			select {
+			case <-p.ctx.Done():
+				p.logger.Infof("exporter handler exiting lastRnd=%d", lastRound)
+				return
+			case pluginBlk := <-blkChan:
+				waitTime := time.Since(selectStart)
+				totalSelectWait += waitTime
+				p.logger.Tracef("exporter handler received block data for round %d after wait of %s", lastRound, waitTime)
+				lastRound = pluginBlk.Round()
+
+				if p.pipelineMetadata.NextRound != lastRound {
+					lastError = fmt.Errorf("aborting after out of order block data. %d != %d", p.pipelineMetadata.NextRound, lastRound)
+					return
+				}
+
+				var exportTime time.Duration
+				exportTime, lastError = RetriesNoOutput(exporter.Receive, pluginBlk.BlockData, p, exporter.Metadata().Name)
+				if lastError != nil {
+					lastError = fmt.Errorf("aborting after failing to export round %d: %w", lastRound, lastError)
+					return
+				}
+				metrics.ExporterTimeSeconds.Observe(exportTime.Seconds())
+				// Ignore round 0 (which is empty).
+				if lastRound > 0 {
+					// Previously we reported time starting after block fetching is complete
+					// through the end of the export operation. Now that each plugin is running
+					// in its own goroutine, report only the time of the export.
+					addMetrics(pluginBlk.BlockData, exportTime)
+				}
+
+				// Increment Round, update metadata
```

Comment:
Since this is happening inside of a go-routine, to keep everything consistent (NextRound should be updated and metadata save before the next exporter is completed) I would need to coordinate with a channel. Do you think this is worthwhile?

Reply:
I'll have to think about this more.

There is a race condition between the exporter finishing and metadata.json being written where the service could be terminated. So the exporter could be called multiple times for the same round. I guess what I'm suggesting could allow the exporter to be called for the previous round. That doesn't sound good. An unbuffered channel could synchronize it a little bit by blocking the exporter from writing results until the callbacks execute, but it would be one round ahead of the callbacks.

Reply:
Yes, callbacks as well are part of the `finishRound()` logic.

Reply:
I think you've convinced me that this should stay here. The timing metadata collected from other plugins still seems out of place, I think you can resolve this thread since we've got another about that already.